File: Python/IPC_HW/hw2.py
#!/usr/bin/env python

import sys

def sum1 (n) :
    (sum, sign) = (n, 1)
    S = ['+', '-']
    sys.stdout.write ("%d" % (n))
    n /= 10
    while n > 0 :
        sys.stdout.write ("%c%d" % (S[sign], n))
        sum += n*(-1)**sign
        sign = 1 - sign
        n /= 10

    sys.stdout.write (" = %d\n" % (sum))

def sum2 (n) :
    (l, sum) = (len (str (n)) - 1, 0)
    while l >= 0 :
        sys.stdout.write ("%d+" % (n / (10**l)))
        sum += (n / (10**l))
        l -= 1

    sys.stdout.write ("\b = %d\n" % (sum))

def sum3 (n) :
    (l, sum) = (len (str (n)), 0)
    for i in range (1, l+1) :
        sys.stdout.write ("%0*d+" % (i, n % (10**i)))
        sum += (n % (10**i))

    sys.stdout.write ("\b = %d\n\n" % (sum))


if __name__ == '__main__' :
    while 1 : 
        try :
            n = int(raw_input ("Enter an integer >=0: ")) 
            sum1 (n)
            sum2 (n)
            sum3 (n)
        except EOFError :
            break

File: Perl/IPC_HW/hw2.pl
#!/usr/bin/perl

use integer;

sub sum1 {
    $n = shift;
    ($sum, $sign) = ($n, 1);
    @S = ('+', '-');
    print "$n";
    $n /= 10;
    while ($n > 0) {
      print "$S[$sign]$n";
      $sum += $n*(-1)**$sign;
      $sign = 1 - $sign;
      $n /= 10;
    }
    print " = $sum\n";
}

sub sum2 {
    $n = shift;
    ($l, $sum) = (length($n) - 1, 0);
    for (reverse 0 .. $l) {
        $sum += ($n / (10**$_));
        print $n / (10**$_) . "+";
    }
    print "\b = $sum\n";
}

sub sum3 {
    $n = shift;
    ($l, $sum) = (length($n), 0);
    for (1 .. $l) {
        $sum += ($n % (10**$_));
        printf "%0*d+", $_, $n % (10**$_);
    }
    print "\b = $sum\n\n";
}

print "Enter an integer >=0: ";
while (chomp($_ = <>)) {
    sum1($_);
    sum2($_);
    sum3($_);
    print "Enter an integer >=0: ";
}
